src/winforms/toga_winforms/widgets/base.py

```diff
@@ -75,11 +78,23 @@ def set_background_color(self, color):
         # By default, background color can't be changed.
         pass
 
+    def winforms_remove_event_handlers(self):
+        for handler in self.winforms_event_handlers:
+            handler['event'] -= handler['handler']
+
     # INTERFACE
 
     def add_child(self, child):
         if self.container:
             child.container = self.container
 
+    def delete_child(self, child):
+        child.winforms_remove_event_handlers()
+        self.native.Controls.Remove(child.native)
+        child.native.Dispose()
```

Comment:
I don't think a separate `delete_child()` method is required. I can't think of a meaningful way that we'd differentiate at an API level between "remove" and "delete".

We *do* need to make sure there's adequate clean up, though. The removal of event handlers and call to Dispose should probably be tied into the `__del__` handler for the impl-level widget, so when the Python garbage collector disposes of the object, those pieces are cleaned up.

That said - I'd be slightly surprised if this isn't already happening. I know the .NET guidelines say that cleanup must be done explicitly, but that's because it has a history with explicit-memory managed languages. It's entirely possible this cleanup & dispose is being done automatically by the python.NET lifecycle.

The test: Write an app that rapidly creates, add and removes children, in a loop. If we see a memory leak over time, and the GC is being invoked, then we'll definitely need to add some `__del__` handling.

Reply:
I've written a [test app](https://gist.github.com/obulat/caddd0d037df92f068deb6ada9c4a6e7) that adds and removes a single button to a container in a loop. What's the best way of testing memory leaks?
Simply running the app seems to show that it does leak: when I run the loop 1000 times, the app runs for about 4 seconds, and when I set it to repeat 100000 times, the app hung.

Reply:
Best method for tracking memory in this situation is probably the resources module: 

```
import resource

usage = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
```

That will tell you the memory allocated to the process.

If 1000x create and destroy takes about four seconds, put that logic on a button handler, and measure (and print) memory usage after every create/destroy cycle completes. If that number goes up with every press, we've got a leak. 

The number won't be _completely_ stable - it might go up and down a little - but if there's a leak, it will go up a lot - and by a reasonably predictable amount each cycle.

Reply:
Apparently, `resource` is Unix specific and doesn't work on Windows.

Reply:
Ok - two alternatives are [tracemalloc](https://docs.python.org/3/library/tracemalloc.html) (this is slower than resource, but gives a more precise memory calculation), or [gc](https://docs.python.org/3/library/gc.html) (which won't give you a literal *memory* count, but will tell you the number of objects that aren't being garbage collected, and which objects are being caught in a reference cycle).

Reply:
I've tried `gc.get_stats`, it returns the same result at the beginning and at the end of the loop. 
This is what I get with `tracemalloc`: 
```
20000 cycles took 33.329090099999995 seconds
[ Top 10 differences ]
<...>\AppData\Local\Programs\Python\Python37\lib\asyncio\windows_events.py:427: size=165 KiB (+165 KiB), count=41 (+41), average=4129 B
<...>\dev\beeware\toga\venv\lib\site-packages\travertino\declaration.py:174: size=95.2 KiB (+95.2 KiB), count=1610 (+1610), average=61 B
<...>\dev\beeware\toga\src\winforms\toga_winforms\app.py:86: size=58.5 KiB (+58.5 KiB), count=667 (+667), average=90 B
<...>\dev\beeware\toga\src\winforms\toga_winforms\app.py:76: size=42.7 KiB (+42.7 KiB), count=471 (+471), average=93 B
<...>\dev\beeware\toga\src\winforms\toga_winforms\window.py:110: size=37.0 KiB (+37.0 KiB), count=442 (+442), average=86 B
<...>\dev\beeware\toga\src\winforms\toga_winforms\libs\proactor.py:103: size=15.2 KiB (+15.2 KiB), count=112 (+112), average=139 B
<...>\dev\beeware\toga\src\winforms\toga_winforms\app.py:83: size=11.4 KiB (+11.4 KiB), count=105 (+105), average=112 B
<...>\dev\beeware\toga\src\winforms\toga_winforms\libs\proactor.py:99: size=8883 B (+8883 B), count=71 (+71), average=125 B
<...>\Local\Programs\Python\Python37\lib\asyncio\windows_events.py:661: size=5936 B (+5936 B), count=44 (+44), average=135 B
<...>\AppData\Local\Programs\Python\Python37\lib\asyncio\windows_events.py:424: size=5904 B (+5904 B), count=41 (+41), average=144 B
```
I'm not sure how to interpret it. [Here is the app code](https://gist.github.com/obulat/caddd0d037df92f068deb6ada9c4a6e7)

Reply:
I've taken a look at this - and it looks like we've got a deeper slow leak. 

Firstly - the really useful method (for our purposes) is [`get_traced_memory()`](https://docs.python.org/3/library/tracemalloc.html#tracemalloc.get_traced_memory). That gives you the current and peak memory allocation of your process (since you made the call to `tracemalloc.start()`). 

Another way to check memory usage is to open the Windows Resource Monitor (Control panel > Administrative tools). That won't log memory to console, but it will give you the OS-level view of how much memory is being used.

Using your test harness as a starting point, and adding in a call to `get_traced_memory()`, I've found some interesting things:

1. Yes, it looks like we've got a leak. I see a significant jump after the 20k iterations.

2. The leak gets a lot worse if you recreate the widget on each iteration. Your test harness creates the dummy button once, and then adds and removes it in rapid succession. If you create a *new* button every time, you see a faster leak.

3. AFAICT, the leak isn't fixed by manually performing the disposal process you've suggested in delete_child. I *think* memory is being cleaned up by the normal Python garbage collector... however, this is a little hard for me to confirm, because...

4. The leak may *not* be related to widget handling at all. If you just leave the app open and do nothing, you'll see in the Resource monitor there's a slow increase in memory usage over time. This is almost certainly related to the asyncio handling; something is accumulating over time.

On that basis, I'm inclined to say we shouldn't worry about the event handling cleanup code for now. We clearly have other memory leaks that are much more significant. Once we've got a handle on the background leak, we can take a closer look at individual widget leaks.

Reply:
Thank you for such detailed answer!

So, for now simply setting container to `None` when removing a child widget is fine, isn't it?

Reply:
I think so, yes.